[Ready to Test] Spritefont styles

Tzenchor opened the pull request:
**As of 20/02/2016 this is ready to test**
I'm working on enabling the style tag of the spritefonts as talked in #4149. At this time i can find the font based on its filename or its family name if it is an installed font in Windows.
The changes at this time include:
- Add 3 private fields to `FontDescriptionProcessor` that contains the supported font extensions, possible terminations of bold font files, and the possible terminations of italic font files.
- Add method `FontDescriptionProcessor.FindFontByFileName`: Considers the fontname as the filename of the font, and tries to find the font file that better matches the name and the style (for example if the users wants "arial" bold, this finds "arialb.ttf"), the method relays heavily on LINQ and I have added comments in each step at least for now, but I don't mind removing them in future.
- Add private class `FontDescriptionProcessor.FileInfoComparer`: Required in method `FontDescriptionProcessor.FindFontByFileName`
- Update method `FontDescriptionProcessor.FindFontFileFromFontName` for Windows: Added style information to the method call and updated its logic.
- Simplified method `FontDescriptionProcessor.Process`: Reworked the method to remove an `if` embedded inside an `#if` and make the procedure simpler.
- Reorder the elements of `FontDescriptionStyle`, now Regular is 0, bold is 1, italic is 2, and by extension bold and italic is 3. Before Bold was 0, italic was 1 and regular was 2.

Both methods look for the best match for the users intentions, giving preference to the italic fonts (if requested).

The code  in `FontDescriptionProcessor.FindFontByFileName` is tested, and linked inside `FontDescriptionProcessor.Process`, but it's result is not used at this time as I want to make the other procedures needed.

I'll need guidance with the following things:
- To search in every folder using family names and to see what styles are available inside each font file i'll need to use SharpFont to read that data, but it feels wrong using it outside `SharpFontImporter`. The problem is that if I do this inside `SharpFontImporter` i'll have to move code from `FontDescriptionProcessor` to `SharpFontImporter`, i'm not sure what option is better.
- Related to the previous point, for rendering the font styles manually I'm gonna need to have information about the found font to know what treatment it needs to get the expected style, and I can do it when retrieving each glyph from the font inside `SharpFontImporter` or after that when back in `FontDescriptionProcessor`
- Also, to review the linux method `FontDescriptionProcessor.FindFontFileFromFontName` i'll need to have an example of the data retrieved in: [line of code](https://github.com/Tzenchor/MonoGame/commit/067476c80767113dd4cd8001b203148a70c649bf#diff-4e98352b3c096b0bf2a04d8f61626bc3L316) from anyone on linux and/or mac

tomspilman commented on MonoGame.Framework.Content.Pipeline/Graphics/FontDescriptionStyle.cs:
```diff
@@ -14,6 +14,11 @@ namespace Microsoft.Xna.Framework.Content.Pipeline.Graphics
         /// <summary>
+        /// Normal text.
+        /// </summary>
+        Regular=0,
```
You should write a unit test verifying this enum is correct to XNA.  Might want to setup a new test specifically for enums in the pipeline sort of like was done here:

https://github.com/mono/MonoGame/blob/develop/Test/Framework/EnumConformingTest.cs

Tzenchor replied:
Test added in the latest commit, i'll need someone to run it in XNA as i don't have the XNA tooling installed.
Also, I have seen that SharpFont includes a similar enum, but with switched values for bold and italic

Tzenchor replied:
Yesterday I got fed up with "we have to compare this with what XNA does", so this morning I have installed the XNA tooling and run the texts to see what happens.
Regarding this enum values I was right, this are the tests result.
![fontdescriptionstyle](https://cloud.githubusercontent.com/assets/11319966/13329327/8a1a15b2-dbf3-11e5-8a39-7cb6135b7aa3.PNG)

tomspilman replied:
Thanks @Tzenchor !

I ran into trouble last year getting the tests running under XNA (see https://github.com/mono/MonoGame/pull/3550).  Now with the next build server setup i'm going to try to get that working soon so that we always test both.

tomspilman commented on MonoGame.Framework.Content.Pipeline/Processors/FontDescriptionProcessor.cs:
```diff
@@ -23,6 +24,18 @@ public class FontDescriptionProcessor : ContentProcessor<FontDescription, Sprite
+        /// <summary>
+        /// Valid font extension
+        /// </summary>
+        String[] _SupportedFontExtensions=new string[] {".ttf",".otf",".ttc" };
```
You should use `string[]` not `String[]`.

tomspilman commented on MonoGame.Framework.Content.Pipeline/Processors/FontDescriptionProcessor.cs:
```diff
@@ -197,19 +203,110 @@ static Glyph[] ImportFont(FontDescription options, out float lineSpacing, out in
+        /// When selecting the file to return, the function returns the better matches the requested style.
+        /// If the requested style is bold and italic, but there's no file that has both styles,
+        /// the first option is returning the italic one, then the bold one, and finally the regular one</returns>
+        string FindFontByFileName(string fontname, List<string> fontDirectories, FontDescriptionStyle style)
```
Generally we skip putting any elaborate documentation on private/internal methods.

Tzenchor replied:
I did it to give the project collaborators a quick overview of what the method should do

tomspilman commented:
Added a couple of initial notes, but I can't look at this in detail until next week.

Tzenchor commented:
I have just added a new method as `FontDescriptionProcessor.FindFontByFontFamily` that looks inside each font file and matches the font's family name with the one from the description, and also looks for multiple styles inside the font file, at least on paper, because I don't have any font that does that.
Also I made the test for `FontDescriptionStyle` and added it to the XNA test suite, but I can't run it

tomspilman commented:
@Tzenchor  -  You need to merge/rebase the latest changes from the `develop` branch back into your `spritefont-styles` feature branch.   This will fix the build process which is failing.

Tzenchor commented:
Ok, will do tomorrow. Also I'll have to rename the test class for the enum, or the report does not work correctly

Tzenchor commented on MonoGame.Framework.Content.Pipeline/Processors/FontDescriptionProcessor.cs:
```diff
@@ -131,7 +136,8 @@ public FontDescriptionProcessor()
 			}
 			catch(Exception ex) {
-				context.Logger.LogImportantMessage("{0}", ex.ToString());
+				//context.Logger.LogImportantMessage("{0}", ex.ToString());
```
Commenting this call i can convert the output to the pipeline from:
![before](https://cloud.githubusercontent.com/assets/11319966/13192705/4e0edcfc-d76f-11e5-8d12-c8a5db546d2c.PNG)
To:
![after](https://cloud.githubusercontent.com/assets/11319966/13192708/52b0fe7a-d76f-11e5-9b0c-19cca655a8db.PNG)
Before merging if there nothing against this change i'll remove the try catch completely

Tzenchor commented:
Once this finishes the building it is ready for testing and review.
I have tried this with all I could and I feel it works well. Note that this tries to return the best match for what the user asks for.
The log messages i added will only show with a debug compilation and with the filter disabled.

Tzenchor commented:
@tomspilman can you rebuild this pull? This is the one that hung the order day

tomspilman commented:
@Tzenchor - Will rebuild it here in a couple of hours...  a complication with the build server is in the way right now.

tomspilman commented:
@Tzenchor - All built...  ignore the "Build Windows 10" failure... that is from the old build setup.  The build is good.

Tzenchor commented:
Ok, thank you.
In that case this is ready for test and review

cra0zy commented:
@Tzenchor You need to resolve merge conflict. Do:
```
git remote add upstream https://github.com/mono/MonoGame.git
git fetch upstream/develop
git rebase upstream/develop
* it will ask you to fix merge conflicts in some files, open them up using texteditor and find what it's talking about, it's gonna have"<<<<<<<<<<<<<<<<<<<<" so you can't miss it.
git add .
git rebase --continue
git push --force
```

Tzenchor commented:
@cra0zy Done, keep in mind that the core of this pull is upgrading the font file finding procedures to take into account the styles of the spritefont

Brett208 commented:
Hello,

I was wondering if there is a way forward on fixing this issue with the SpriteFont Style element issue?

The styles issue hung me up for a while when updating my game from XNA to MonoGame.

-Brett

Tzenchor commented:
Once i am told that this will be reviewed i'll resolve the conflicts, but they are busy with more urgent things

cra0zy commented:
@tomspilman @KonajuGames @dellis1972  I think one of you should take some time to review this.

anetegithub commented:
Any news about this?

wldevries commented:
There is no need to deduce the style from the font file name, because the TrueType / OpenType file format is documented pretty well. Using a managed library like https://github.com/roy-t/TrueType or https://github.com/LayoutFarm/Typography you can parse all installed TrueType font files well within a second even if MS Office is installed.

The `name` table includes Font Family name (e.g. `Arial`) and Font Subfamily name (e.g. `Bold`) for the file

TrueType reference guide (Apple): https://developer.apple.com/fonts/TrueType-Reference-Manual/
OpenType specification (Microsoft): https://learn.microsoft.com/en-us/typography/opentype/spec/

tomspilman commented:
> There is no need to deduce the style from the font file name,

@wldevries  is correct.

For this reason and it being conflicted i'm closing this.  If we fix it and resolve conflicts it can be reopened.
